File: docs/code-examples/depth_image_3d.py
"""Create and log a depth image and pinhole camera."""

import numpy as np
import rerun as rr

depth_image = 65535 * np.ones((8, 12), dtype=np.uint16)
depth_image[0:4, 0:6] = 20000
depth_image[4:8, 6:12] = 45000

rr.init("rerun_example_depth_image", spawn=True)

# If we log a pinhole camera model, the depth gets automatically back-projected to 3D
rr.log(
    "world/camera",
    rr.Pinhole(
        width=depth_image.shape[1],
        height=depth_image.shape[0],
        focal_length=20,
    ),
)

# Log the tensor.
rr.log("world/camera/depth", rr.DepthImage(depth_image, meter=10_000.0))

File: docs/code-examples/depth_image_3d.cpp
// Create and log a depth image.

#include <rerun.hpp>

#include <algorithm>

int main() {
    auto rec = rerun::RecordingStream("rerun_example_depth_image");
    rec.connect().throw_on_failure();

    // Create a synthetic depth image.
    const int HEIGHT = 200;
    const int WIDTH = 300;
    std::vector<uint16_t> data(WIDTH * HEIGHT, 65535);
    for (auto y = 50; y < 150; ++y) {
        std::fill_n(data.begin() + y * WIDTH + 50, 100, 20000);
    }
    for (auto y = 130; y < 180; ++y) {
        std::fill_n(data.begin() + y * WIDTH + 100, 180, 45000);
    }

    // If we log a pinhole camera model, the depth gets automatically back-projected to 3D
    rec.log(
        "world/camera",
        rerun::Pinhole::focal_length_and_resolution(
            {20.0f, 20.0f},
            {static_cast<float>(WIDTH), static_cast<float>(HEIGHT)}
        )
    );

    rec.log(
        "world/camera/depth",
        rerun::DepthImage({HEIGHT, WIDTH}, std::move(data)).with_meter(10000.0)
    );
}
